File: back_end/services/message.services.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/Abhishekh669/backend/configuration"
	"github.com/Abhishekh669/backend/model"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func reverseMessages(messages []map[string]interface{}) {
	for i, j := 0, len(messages)-1; i < j; i, j = i+1, j-1 {
		messages[i], messages[j] = messages[j], messages[i]
	}
}
func populateMessageDetails(message model.Message) (bson.M, error) {
	// Retrieve the sender and receiver IDs

	// Fetch sender and receiver details (assuming you have a function to get user details)
	sender, err := GetUserById(message.Sender)
	if err != nil {
		return nil, fmt.Errorf("failed to get sender details: %v", err)
	}

	receiver, err := GetUserById(message.Receiver)
	if err != nil {
		return nil, fmt.Errorf("failed to get receiver details: %v", err)
	}

	// Handle the seenBy field (if available)

	var seenByUsers []map[string]interface{}
	for _, userID := range message.SeenBy {
		user, err := GetUserById(userID)
		if err != nil {
			return nil, fmt.Errorf("failed to fetch seen_by user: %v", err)
		}
		data := map[string]interface{}{
			"_id":      user.ID,
			"name":     user.Name,
			"email":    user.Email,
			"address":  user.Address,
			"field":    user.Field,
			"image":    user.Image,
			"codeName": user.CodeName,
		}
		seenByUsers = append(seenByUsers, data)
	}

	// Populated message structure
	populatedMessage := bson.M{
		"_id":            message.ID,
		"conversationId": message.ConversationId,
		"body":           message.Body,
		"sender": map[string]interface{}{
			"_id":      sender.ID,
			"name":     sender.Name,
			"email":    sender.Email,
			"address":  sender.Address,
			"field":    sender.Field,
			"image":    sender.Image,
			"codeName": sender.CodeName,
		},
		"receiver": map[string]interface{}{
			"_id":      receiver.ID,
			"name":     receiver.Name,
			"email":    receiver.Email,
			"address":  receiver.Address,
			"field":    receiver.Field,
			"image":    receiver.Image,
			"codeName": receiver.CodeName,
		},
		"seenBy":    seenByUsers,
		"createdAt": message.CreatedAt, // Here, it's an array of strings instead of ObjectIDs
	}

	return populatedMessage, nil
}

func populatedMessagesDetails(message bson.M) (bson.M, error) {
	senderID, ok := message["sender"].(primitive.ObjectID)
	if !ok {
		return nil, fmt.Errorf("invalid sender ID")
	}

	receiverID, ok := message["receiver"].(primitive.ObjectID)
	if !ok {
		return nil, fmt.Errorf(("invlaid reciever ID"))
	}

	sender, err := GetUserById(senderID)
	if err != nil {
		return nil, fmt.Errorf("failed to get sender: %v", err)
	}

	receiver, err := GetUserById(receiverID)
	if err != nil {
		return nil, fmt.Errorf("failed to get receiver: %v", err)
	}

	seenByIDs, ok := message["seenBy"].(bson.A)
	if !ok {
		return nil, fmt.Errorf("invalid seenBy field")
	}

	var seenByUsers []map[string]interface{}
	for _, userID := range seenByIDs {
		user, err := GetUserById(userID.(primitive.ObjectID))
		if err != nil {
			return nil, fmt.Errorf("failed to fetch seen_by user: %v", err)
		}
		data := map[string]interface{}{
			"_id":      user.ID,
			"name":     user.Name,
			"email":    user.Email,
			"address":  user.Address,
			"field":    user.Field,
			"image":    user.Image,
			"codeName": user.CodeName,
		}
		seenByUsers = append(seenByUsers, data)
	}

	populatedMessage := bson.M{
		"_id":            message["_id"],
		"conversationId": message["conversationId"],
		"body":           message["body"],
		"sender": map[string]interface{}{
			"_id":      sender.ID,
			"name":     sender.Name,
			"email":    sender.Email,
			"address":  sender.Address,
			"field":    sender.Field,
			"image":    sender.Image,
			"codeName": sender.CodeName,
		},
		"receiver": map[string]interface{}{
			"_id":      receiver.ID,
			"name":     receiver.Name,
			"email":    receiver.Email,
			"address":  receiver.Address,
			"field":    receiver.Field,
			"image":    receiver.Image,
			"codeName": receiver.CodeName,
		},

		"seenBy": seenByUsers,
	}
	return populatedMessage, nil
}

func GetAllConversationMessages(id primitive.ObjectID, lastID primitive.ObjectID, limit int) (map[string]interface{}, error) {

	message_collection, err := configuration.GetCollection("messages")
	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("could not get the message collection: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var messages []map[string]interface{}

	filter := bson.M{
		"conversationId": id,
	}

	count, err := message_collection.CountDocuments(ctx, filter)
	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("failed to count messages : %v", err)
	}

	if !lastID.IsZero() {
		filter["_id"] = bson.M{"$lt": lastID} // Fetch messages with _id less than lastID
	}

	options := options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}}).SetLimit(int64(limit))
	cursor, err := message_collection.Find(ctx, filter, options)
	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("failed to get messages: %v", err)
	}
	defer cursor.Close(ctx)

	for cursor.Next(ctx) {
		var message bson.M
		if err := cursor.Decode(&message); err != nil {
			return map[string]interface{}{}, fmt.Errorf("failed to decode message: %v", err)
		}

		populatedMessage, err := populatedMessagesDetails(message)
		if err != nil {
			return map[string]interface{}{}, fmt.Errorf("failed to populate message details: %v", err)
		}

		messages = append(messages, populatedMessage)
	}

	if err := cursor.Err(); err != nil {
		return map[string]interface{}{}, fmt.Errorf("cursor error: %v", err)
	}

	reverseMessages(messages)

	response := map[string]interface{}{
		"messages":      messages,
		"totalMessages": count,
	}

	return response, nil
}

func CreateMesssage(message model.Message) (map[string]interface{}, error) {
	message_collection, err := configuration.GetCollection("messages")

	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("could not get the message collection: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)

	defer cancel()

	result, err := message_collection.InsertOne(ctx, message)

	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("failed to create message: %v", err)
	}

	message.ID = result.InsertedID.(primitive.ObjectID)

	populatedMessage, err := populateMessageDetails(message)
	if err != nil {
		return map[string]interface{}{}, fmt.Errorf("failed to populate message details: %v", err)
	}
	return populatedMessage, nil
}

func GetMessageById(id primitive.ObjectID) (model.Message, error) {

	var message model.Message
	filter := bson.D{{Key: "_id", Value: id}}

	message_collection, err := configuration.GetCollection("messages")

	if err != nil {
		return model.Message{}, fmt.Errorf("could not get the message collection: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	err = message_collection.FindOne(ctx, filter).Decode(&message)
	fmt.Println("THis ishte reuslt after search the message ", message)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			return model.Message{}, fmt.Errorf("no message found : %v", err)
		} else {
			return model.Message{}, fmt.Errorf("failed to find the message")
		}
	}
	fmt.Println("THis is check user ", message)
	return message, nil
}

File: chat_back_end/src/services/message.services.ts
import { ObjectId, WithId } from "mongodb";
import { getCollection } from "../lib/db/db";
import { MessageEditModel, MessageModel } from "../models/messages.model";




const populateMessageDetails = async (message: MessageModel) => {
  const senderData = await getUserById(message.senderId);
  if (!senderData || !senderData._id || !senderData.email) {
    return null;
  }
  let receiverData;
 if(message.receiverId){
   receiverData = await getUserById(message.receiverId);
  if (!receiverData || !receiverData._id || !receiverData.email) {
    return null;
  }
 }

  var SeenByUsers = message.seenBy
  ? await Promise.all(
      message.seenBy.map(async (userId: ObjectId) => {
        const user = await getUserById(userId);
        return user
          ? {
              _id: user._id,
              name: user.name,
              email: user.email,
              address: user.address,
              field: user.field,
              image: user.image,
              codeName: user.codeName,
            }
          : null;
      })
    )
  : [];

  const populatedMessage = {
    _id: message._id,
    conversationId: message.conversationId,
    text: message.text,
    sender: {
      _id: senderData._id,
      name: senderData.name,
      email: senderData.email,
      address: senderData.address,
      field: senderData.field,
      image: senderData.image,
      codeName: senderData.codeName,
    },
    receiver: receiverData ? {
      _id: receiverData._id,
      name: receiverData.name,
      email: receiverData.email,
      address: receiverData.address,
      field: receiverData.field,
      image: receiverData.image,
      codeName: receiverData.codeName,
    } : null,
    seenBy : SeenByUsers,
    ...(message.replyTo && {replyTo : message.replyTo}),
    ...(message.reactions && {reactions : message.reactions}),
    createdAt : message.createdAt,
    updatedAt : message.updatedAt,
  };
  return populatedMessage
};



export const  deleteMessageById = async(id : ObjectId) =>{
  try {
    const messageCollection = await getCollection("messages");
    const deleteResult = await messageCollection.deleteOne({ _id: id });
    if (deleteResult.deletedCount === 0) return null;
    return true;
  } catch (error) {
    return null;
  }

}
  



export const getMessageById = async (id : ObjectId) =>{
  try {
    const messageCollection = await getCollection("messages")
    const findMessage = await messageCollection.findOne({_id : id})
    if(!findMessage)return null;
    return findMessage;
    
  } catch (error) {
    return null;
    
  }
}



export const EditMessage = async(message : MessageEditModel)=>{
  try {
    if(!message._id || !message.conversationId ){
      return null;
    }

    const messageCollection = await getCollection("messages")
    const result = await messageCollection.updateOne(
      {
        _id : message._id,
        conversationId : message.conversationId,
      },
      {
        $set : {
          text : message.text,
          updatedAt : new Date(),
        }
      }
    )

    if(result.matchedCount === 0){
      return null;
    }
    
    const doc = await getMessageById(message._id)
    if(!doc)return null;
    const messageModelMessage = {
      _id: doc._id,
      senderId: doc.senderId,
      receiverId: doc.receiverId,
      text: doc.text,
      image: doc.image,
      conversationId: doc.conversationId,
      createdAt: doc.createdAt,
      updatedAt : doc.updatedAt,
      seenBy: doc.seenBy,
      reactions: doc.reactions,
      replyTo: doc.replyTo,
    }
    const populatedMessage = await populateMessageDetails(messageModelMessage)
    return populatedMessage;
      
  } catch (error) {
    return null;
    
  }
  

}


export const createNewMessage = async (
  messageData: Omit<MessageModel, "_id">
) => {
  try {
    if (!messageData) return null;
    const messages = await getCollection("messages");
    const insertMessages = await messages.insertOne(messageData);
    if (!insertMessages.insertedId) {
      return null;
    }
    const populatedMessageDetails = await populateMessageDetails({_id : insertMessages.insertedId, ...messageData})
    if(!populatedMessageDetails)return null;
    return populatedMessageDetails;
  } catch (error) {
    return null;
  }
};

export const getUserById = async (userId: ObjectId) => {
  const userCollection = await getCollection("users");
  return userCollection.findOne({ _id: userId });
};

export const getAllConversationMessages = async (
  conversationId: string,
  lastId: string,
  limit: number
) => {
  try {
    const messageCollection = await getCollection("messages");
    if (!messageCollection) {
      return null;
    }
    const filter: any = { conversationId: new ObjectId(conversationId) };

    const count = await messageCollection.countDocuments(filter);

    if (lastId) {
      filter._id = { $lt: new ObjectId(lastId) };
    }

    const messages = await messageCollection
      .find(filter)
      .sort({ createdAt: -1 })
      .limit(Number(limit) || 10)
      .toArray();

    if (!messages) {
      return null;
    }
  } catch (error) {
    return null;
  }
};
